internal/walk/walklinks.go

```diff
@@ -31,6 +31,28 @@ func walkDirWithLinks(ctx context.Context, root string, fn fs.WalkDirFunc) error
 			return ctx.Err()
 		}
 
+		if len(exclude) > 0 {
+			for _, ex := range exclude {
+				var inExcluded bool
+
+				// Special handling for globs to publish the parent dir and exclude its children.
+				if strings.HasSuffix(ex, "/*") {
+					inExcluded = strings.HasSuffix(filepath.Dir(path), strings.TrimRight(ex, "/*"))
+				}
+
+				isExcluded, err := filepath.Match(ex, d.Name())
```

Comment:
I think this is overly simplistic and won't work in some important cases.

A case I have in mind right now is "content/contentisos", which is supposed to match a directory, for RHELDST-6189.  d.Name() here is only going to be a single path component, e.g. "content" or "contentisos", and will never match the exclusion pattern in the same way as rsync does.

I see two issues which might need to be addressed here:

1. rsync matches the pattern against every component of the path, but this code is only matching against one component.  From "man rsync":

> Note that, when using the --recursive (-r) option (which is implied by -a), every subdir component of
> every  path  is  visited  left to right, with each directory having a chance for exclusion before its
> content [...]
> For instance, to include "/foo/bar/baz", the directories "/foo" and "/foo/bar" must not be  excluded.

I understand we're doing this match at every level of the walk, but still I don't think you can get compatible behavior with rsync in this way.  e.g. with the above example, rsync could understand an exclusion pattern of "foo/bar" but this code wouldn't be able to handle it.

It seems necessary to match against the full path here, i.e. `path + "/" + d.Name()` .

2. filepath.Match says that the pattern has to match the entire input.  Unfortunately it seems that'll complicate the fixing of (1) because if you literally just change this to test against the whole path, it won't work - it'll need some other changes as well.

I'm not sure this can be done by just one call to filepath.Match, it seems to me that something more complicated might be needed.

Considering example scenario:

- `path == "/my/tmpdir/content"`
- `d.Name() == "contentisos"`
- `ex == "content/contentisos"`

I have two suggestions how to change the logic here:

### Use regex

Scrap the idea of using filepath.Match.  Instead, convert the input pattern into a regular expression and then compare it over the whole path.  This would also fix several other incompatibilities with rsync, such as lack of support for `**`.

I'm looking at "man rsync", "INCLUDE/EXCLUDE PATTERN RULES", and it seems like many of the behaviors there can be translated into regular expression syntax fairly well, for instance:

- `/` at beginning of pattern => `^/`
- `**` => `.*`
- `*` => `[^\]+`
- `?` => `[^\]`

In this example, "content/contentisos" converted to a regular expression is just "content/contentisos" (no changes at all).  When matched against full path "/my/tmpdir/content/contentisos", it matches the regex and so excludes it.

### Match path components from right-to-left

If you want to stick with filepath.Match, a way to cope with the behavior that it "requires pattern to match all of name" would be to build up partial components of the path from right to left and test each one against the pattern.

The algorithm would work out like this:

- assemble full path: "/my/tmpdir/content/contentisos"
- loop over partial paths starting from rightmost component and adding one more leading component at each iteration:
   - "contentisos" => no match
   - "content/contentisos" => MATCH, excluded.
   - if it hadn't already matched, next paths checked would be:
      - "tmpdir/content/contentisos"
      - "my/tmpdir/content/contentisos"
      - "/my/tmpdir/content/contentisos" (if ex started with "/", it'd finally have a chance to match here)
      - stop.

This still leaves some obvious incompatibilities with rsync such as handling of `**`, but I can't think of any genuine expected usage where it fails.